### ISSUE TITLE
[0.73 RC.4][iOS] Multiple jsi.h from different sources

### ISSUE THREAD

**Issue by Kudo** [facebook/react-native#41498]
### New Version

0.73.0-rc.4

### Old Version

0.72.5

### Build Target(s)

iOS

### Output of `react-native info`
```
System:
  OS: macOS 14.0
  CPU: (8) arm64 Apple M1 Pro
  Memory: 686.64 MB / 32.00 GB
  Shell:
    version: 3.6.1
    path: /opt/homebrew/bin/fish
Binaries:
  Node:
    version: 18.18.0
    path: /var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/yarn--1700049422150-0.40821589002556014/node
  Yarn:
    version: 1.22.19
    path: /var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/yarn--1700049422150-0.40821589002556014/yarn
  npm:
    version: 9.8.1
    path: ~/.volta/tools/image/node/18.18.0/bin/npm
  Watchman:
    version: 2023.10.02.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.12.1
    path: /Users/kudo/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.0
      - iOS 17.0
      - macOS 14.0
      - tvOS 17.0
      - watchOS 10.0
  Android SDK:
    API Levels:
      - "31"
      - "33"
      - "34"
    Build Tools:
      - 30.0.3
      - 31.0.0
      - 33.0.0
      - 33.0.1
      - 34.0.0
    System Images:
      - android-31 | Google Play ARM 64 v8a
      - android-33 | Google APIs ARM 64 v8a
      - android-33 | Google Play ARM 64 v8a
      - android-34 | Google Play ARM 64 v8a
    Android NDK: 23.1.7779620
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10406996
  Xcode:
    version: 15.0.1/15A507
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.6
    path: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/javac
  Ruby:
    version: 2.7.4
    path: /Users/kudo/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.0-rc.4
    wanted: 0.73.0-rc.4
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Issue and Reproduction Steps

There is a regression from 0.73.0-rc.4 where it contains multiple jsi.h, one from React-jsi and the other from hermes-engine.

```
ios/Pods/Headers/Public/React-jsi/jsi/jsi.h
ios/Pods/Headers/Public/hermes-engine/jsi/jsi.h
ios/Pods/Headers/Private/React-jsi/jsi/jsi.h
ios/Pods/Headers/Private/hermes-engine/jsi/jsi.h
ios/Pods/hermes-engine/destroot/include/jsi/jsi.h
```

Since the jsi.h have different paths, they will cause unpredictable resolutions. At Expo, we have some test cases to throw `jsi::JSError` from hermes-engine and catch the exceptions out of hermes-engine. In hermes-engine, it uses the jsi.h from hermes-engine. On the other hand, on app side it uses the jsi.h from React-jsi to catch the jsi::JSError. Then the exceptions are not uncaught and make app crash.

The relevant change seems to be https://github.com/facebook/react-native/commit/a66ebc5d97fe21c561ec1c448f2532d79001f31c

## Reproduce steps

```sh
$ npx react-native@latest init RN073 --version next
$ cd RN073
$ find ios/Pods -name 'jsi.h'
ios/Pods/Headers/Public/React-jsi/jsi/jsi.h
ios/Pods/Headers/Public/hermes-engine/jsi/jsi.h
ios/Pods/Headers/Private/React-jsi/jsi/jsi.h
ios/Pods/Headers/Private/hermes-engine/jsi/jsi.h
ios/Pods/hermes-engine/destroot/include/jsi/jsi.h
```

**Comment by dmytrorykun** [comment]
Pick requested https://github.com/reactwg/react-native-releases/discussions/64#discussioncomment-7587141

**Comment by Kudo** [comment]
can confirm this is resolved by #41499. thanks @dmytrorykun and @cipolleschi
